### PR TITLE
Introduce `AdjointFactorization` not subtyping `AbstractMatrix`

### PR DESCRIPTION
**Pull request by dkarrasch**
I have long been bugged by the fact that, on the one hand, you can wrap anything by `Adjoint`, even types that don't subtype to `AbstractMatrix`, but once wrapped, they are back in the `AbstractMatrix` hierarchy. It turns out that this subtyping basically doesn't lead to any usage of fallbacks, except for `size(::Adjoint{Factorization}, ::Int)`! ~~So, this disentangles that, and even restricts `Adjoint`/`Transpose` to wrap only `AbstractVecOrMat`, which is not necessary. Anyway, it shows that it can be done.~~ Obviously, this is breaking, but I assume it breaks only a few basic packages, but we'll see in a nanosoldier run.

This is a "companion" PR to #46196 and #46233.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/factorization.jl [hunk 1]
```diff
@@ -60,64 +83,76 @@ convert(::Type{T}, f::Factorization) where {T<:AbstractArray} = T(f)::T
 
 ### General promotion rules
 Factorization{T}(F::Factorization{T}) where {T} = F
-# This is a bit odd since the return is not a Factorization but it works well in generic code
-Factorization{T}(A::Adjoint{<:Any,<:Factorization}) where {T} =
+# This no longer looks odd since the return _is_ a Factorization!
+Factorization{T}(A::AdjointFactorization) where {T} =
     adjoint(Factorization{T}(parent(A)))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/factorization.jl:88]: 😉

**Comment by dkarrasch** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by dkarrasch** [comment]
I guess if we don't restrict `Adjoint` to `AbsVecOrMat`, then this is actually "not"  breaking (unless people define functions for `Adjoint{<:Any,<:F}` where `F` is a subtype to some factorization from `LinearAlgebra`). I see usage as in `GenericLinearAlgebra` for `Householder` and `HouseholderBlock`, which both don't subtype to `AbstractMatrix`, even though their behavior is "matrix-like", e.g., `size(H, i)` returns 1 if `i > 2`. If we wanted to restrict `Adjoint` to `AbsVecOrMat`, what would we do with `Householder[Block]`? (1) Introduce a minimal `AdjointHouseholder` type? (2) Subtype `Householder <: AbstractMatrix`? Or is that an indication that we shouldn't add the restriction? Do we care that `Adoint{Householder}` is an `AbstractMatrix`? @andreasnoack

**Comment by dkarrasch** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46874#issuecomment-1299934965) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/5683e7d_vs_df06375/report.html).

**Comment by dkarrasch** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46874#issuecomment-1424268085) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e88cc96_vs_7caa499/report.html).

**Comment by dkarrasch** [comment]
I believe the ecosystem is ready for this. Shall we wait until all PRs are merged and run another pkgeval test or can we merge this right away?

**Comment by dkarrasch** [comment]
Perhaps this needs a NEWS entry.

**Comment by dkarrasch** [comment]
@nanosoldier `runtests(["SparseMatricesCSR", "RecursiveFactorization", "KLU", "UpdatableQRFactorizations", "SuperLU", "Juniper", "Infernal", "QuantumTomography", "BandedMatrices"])`

**Comment by nanosoldier** [comment]
[Your job](https://github.com/JuliaLang/julia/pull/46874#issuecomment-1429441200) failed. Consult the server logs for more details (cc @maleadt).

**Comment by dkarrasch** [comment]
@nanosoldier `runtests(["SparseMatricesCSR", "RecursiveFactorization", "KLU", "UpdatableQRFactorizations", "SuperLU", "Juniper", "Infernal", "QuantumTomography", "BandedMatrices"])`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46874#issuecomment-1431294996) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b31311a_vs_cd955e1/report.html).

**Comment by dkarrasch** [comment]
The remaining failing packages have open PRs, so I think this is finally ready to go?

**Comment by dkarrasch** [comment]
The only failing test is unrelated. Let's go!

**Pull request merged**